### PR TITLE
fix: fixes workspace creation with a remote backend.

### LINKED ISSUE #824: Tofu client should send terraform-version (or tofu-version) when creating new workspace
### OpenTofu Version

```shell
OpenTofu v1.6.0-alpha3
on linux_arm64
```

### OpenTofu Configuration Files

terraform {
  backend "remote" {
     hostname = "<hostname to private terraform registry>"
     organization = "<organization in private terraform registry>""
     workspaces {
         name = "test"
     }
  }
}

### Debug Output

2023-11-06T10:31:22.466Z [INFO]  OpenTofu version: 1.6.0 alpha3
       2023-11-06T10:31:22.468Z [DEBUG] using github.com/hashicorp/go-tfe v1.32.1
       2023-11-06T10:31:22.468Z [DEBUG] using github.com/hashicorp/hcl/v2 v2.18.1
       2023-11-06T10:31:22.468Z [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
       2023-11-06T10:31:22.468Z [DEBUG] using github.com/zclconf/go-cty v1.14.1
       2023-11-06T10:31:22.468Z [INFO]  Go runtime version: go1.21.3
       2023-11-06T10:31:22.468Z [INFO]  CLI args: []string{"tofu", "init"}
       2023-11-06T10:31:22.468Z [TRACE] Stdout is not a terminal
       2023-11-06T10:31:22.468Z [TRACE] Stderr is not a terminal
       2023-11-06T10:31:22.468Z [TRACE] Stdin is not a terminal
       2023-11-06T10:31:22.468Z [DEBUG] Attempting to open CLI config file: /root/.tofurc
       2023-11-06T10:31:22.468Z [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
       2023-11-06T10:31:22.468Z [INFO]  Loading CLI configuration from /root/.terraform.d/credentials.tfrc.json
       2023-11-06T10:31:22.468Z [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
       2023-11-06T10:31:22.468Z [DEBUG] ignoring non-existing provider search directory /root/.terraform.d/plugins
       2023-11-06T10:31:22.468Z [DEBUG] ignoring non-existing provider search directory /root/.local/share/terraform/plugins
       2023-11-06T10:31:22.468Z [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
       2023-11-06T10:31:22.468Z [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
       2023-11-06T10:31:22.469Z [DEBUG] Found the config directory: /root/.terraform.d
       2023-11-06T10:31:22.469Z [INFO]  CLI command args: []string{"init"}

       Initializing the backend...
       2023-11-06T10:31:22.469Z [TRACE] Meta.Backend: built configuration for "remote" backend with hash value 2792484100
       2023-11-06T10:31:22.469Z [TRACE] Meta.Backend: backend has not previously been initialized in this working directory
       2023-11-06T10:31:22.469Z [DEBUG] New state was assigned lineage "5f70810e-357c-c542-c2e6-3b3d0f85ce52"
       2023-11-06T10:31:22.469Z [TRACE] Meta.Backend: moving from default local state only to "remote" backend
       2023-11-06T10:31:22.469Z [DEBUG] checking for provisioner in "."
       2023-11-06T10:31:22.470Z [DEBUG] checking for provisioner in "/usr/local/bin"
       2023-11-06T10:31:22.470Z [TRACE] backend/local: state manager for workspace "default" will:
        - read initial snapshot from terraform.tfstate
        - write new snapshots to terraform.tfstate
        - create any backup at terraform.tfstate.backup
       2023-11-06T10:31:22.470Z [TRACE] statemgr.Filesystem: reading initial snapshot from terraform.tfstate
       2023-11-06T10:31:22.470Z [TRACE] statemgr.Filesystem: snapshot file has nil snapshot, but that's okay
       2023-11-06T10:31:22.470Z [TRACE] statemgr.Filesystem: read nil snapshot
       2023-11-06T10:31:22.470Z [TRACE] Meta.Backend: ignoring local "default" workspace because its state is empty
       2023-11-06T10:31:22.470Z [DEBUG] Service discovery for tflocal.com at https://tflocal.com/.well-known/terraform.json
       2023-11-06T10:31:22.470Z [TRACE] HTTP client GET request to https://tflocal.com/.well-known/terraform.json
       2023-11-06T10:31:22.547Z [TRACE] Meta.BackendForPlan: backend *remote.Remote supports operations
       2023-11-06T10:31:22.548Z [DEBUG] Service discovery for tflocal.com aliased as localterraform.com
       2023-11-06T10:31:22.548Z [DEBUG] New state was assigned lineage "586927fe-ceb3-e127-3d15-d6fcfc239a2b"
       2023-11-06T10:31:22.562Z [TRACE] Preserving existing state lineage "586927fe-ceb3-e127-3d15-d6fcfc239a2b"

      
       Successfully configured the backend "remote"! OpenTofu will automatically
       use this backend unless the backend configuration changes.
        Error: No existing workspaces.
     
      Use the "tofu workspace" command to create and select a new workspace.
       If the backend already contains existing workspaces, you may need to update
      the backend configuration.
      
       2023-11-06T10:31:22.581Z [INFO]  OpenTofu version: 1.6.0 alpha3
       2023-11-06T10:31:22.581Z [DEBUG] using github.com/hashicorp/go-tfe v1.32.1
       2023-11-06T10:31:22.581Z [DEBUG] using github.com/hashicorp/hcl/v2 v2.18.1
       2023-11-06T10:31:22.581Z [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
       2023-11-06T10:31:22.581Z [DEBUG] using github.com/zclconf/go-cty v1.14.1
       2023-11-06T10:31:22.581Z [INFO]  Go runtime version: go1.21.3
       2023-11-06T10:31:22.581Z [INFO]  CLI args: []string{"tofu", "workspace", "new", "test1"}
       2023-11-06T10:31:22.581Z [TRACE] Stdout is not a terminal
       2023-11-06T10:31:22.581Z [TRACE] Stderr is not a terminal
       2023-11-06T10:31:22.581Z [TRACE] Stdin is not a terminal
       2023-11-06T10:31:22.581Z [DEBUG] Attempting to open CLI config file: /root/.tofurc
       2023-11-06T10:31:22.581Z [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
       2023-11-06T10:31:22.581Z [INFO]  Loading CLI configuration from /root/.terraform.d/credentials.tfrc.json
       2023-11-06T10:31:22.581Z [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
       2023-11-06T10:31:22.581Z [DEBUG] ignoring non-existing provider search directory /root/.terraform.d/plugins
       2023-11-06T10:31:22.581Z [DEBUG] ignoring non-existing provider search directory /root/.local/share/terraform/plugins
       2023-11-06T10:31:22.581Z [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
       2023-11-06T10:31:22.581Z [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
       2023-11-06T10:31:22.581Z [DEBUG] Found the config directory: /root/.terraform.d
       2023-11-06T10:31:22.581Z [INFO]  CLI command args: []string{"workspace", "new", "test1"}
       2023-11-06T10:31:22.582Z [TRACE] Meta.Backend: built configuration for "remote" backend with hash value 2792484100
       2023-11-06T10:31:22.582Z [TRACE] Preserving existing state lineage "586927fe-ceb3-e127-3d15-d6fcfc239a2b"
       2023-11-06T10:31:22.582Z [TRACE] Preserving existing state lineage "586927fe-ceb3-e127-3d15-d6fcfc239a2b"
       2023-11-06T10:31:22.582Z [TRACE] Meta.Backend: working directory was previously initialized for "remote" backend
       2023-11-06T10:31:22.582Z [TRACE] Meta.Backend: using already-initialized, unchanged "remote" backend configuration
       2023-11-06T10:31:22.582Z [DEBUG] Service discovery for tflocal.com at https://tflocal.com/.well-known/terraform.json
       2023-11-06T10:31:22.582Z [TRACE] HTTP client GET request to https://tflocal.com/.well-known/terraform.json
       2023-11-06T10:31:22.618Z [TRACE] Meta.BackendForPlan: backend *remote.Remote supports operations
       2023-11-06T10:31:22.618Z [DEBUG] Service discovery for tflocal.com aliased as localterraform.com
       2023-11-06T10:31:22.618Z [TRACE] Meta.Backend: instantiated backend of type *remote.Remote
       2023-11-06T10:31:22.618Z [DEBUG] checking for provisioner in "."
       2023-11-06T10:31:22.618Z [DEBUG] checking for provisioner in "/usr/local/bin"
       2023-11-06T10:31:22.618Z [TRACE] Meta.Backend: backend *remote.Remote supports operations
       Remote workspace OpenTofu version "" does not match local OpenTofu version "1.6.0-alpha3"

### Expected Behavior

tofu should be able to create a workspace

### Actual Behavior

Tofu failing to create a workspace using tofu workspace new command

### Steps to Reproduce

tofu init
tofu workspace new test

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by kislerdm**
Resolves #824

## Target Release

1.6.0

### BUG FIXES

- Fixes workspace creation for remote backend

### ENHANCEMENTS

- Adds the local OpenTofu version upon creation of the workspace in the case of remove backend
- Removes comparison of the local OpenTofu version to the remote backend workspace's version

### CODE REVIEW

File: internal/backend/remote/backend.go [hunk 1]
```diff
@@ -679,20 +675,6 @@ func (b *Remote) StateMgr(ctx context.Context, name string) (statemgr.Full, erro
...
-	// accidentally upgrade state with a new code path, and the version check
-	// logic is coarser and simpler.
-	if !b.ignoreVersionConflict {
-		wsv := workspace.TerraformVersion
-		// Explicitly ignore the pseudo-version "latest" here, as it will cause
-		// plan and apply to always fail.
-		if wsv != tfversion.String() && wsv != "latest" {
-			return nil, fmt.Errorf("Remote workspace OpenTofu version %q does not match local OpenTofu version %q", workspace.TerraformVersion, tfversion.String())
-		}
-	}
```
Comment (Yantrio) [inline internal/backend/remote/backend.go:694]: I still think we want this check, we don't want someone mixing versions up. 

in this case i think we should still throw an error if someone has a workspace version of 1.6 and a tfversion of 1.9 (for example).

Comment (cam72cam) [reply internal/backend/remote/backend.go:694]: @kislerdm and I are on a call right now puzzling out where "latest" is supposedly set.

Comment (cam72cam) [reply internal/backend/remote/backend.go:694]: Our hypothesis is that the remote workspace is supposed to default to "latest" when terraform-version is not set.  JFrog's Artifactory probably does not include this implementation Quirk.

Comment (cam72cam) [reply internal/backend/remote/backend.go:694]: There are other overrides for version = "latest" in other parts of the workspace code that @kislerdm identified

Comment (cam72cam) [reply internal/backend/remote/backend.go:694]: IMO terraform has weird hacks around alphas, which is what is biting us at the moment.

Comment (kislerdm) [reply internal/backend/remote/backend.go]: @Yantrio updated. Please have a look.

File: internal/backend/remote/backend.go [hunk 2]
```diff
@@ -716,6 +709,28 @@ func (b *Remote) StateMgr(ctx context.Context, name string) (statemgr.Full, erro
...
 }
 
+func localVersionMatchesRemoteVersion(wsv string) error {
+	const latestVersion = "latest"
+
+	if wsv == latestVersion {
+		// We handle the "latest" for compatibility reasons:
+		// TFE and potentially other backends can set the workspace version to "latest"
+		// if they were created using pre-released version of terraform
+		return nil
```
Comment (kislerdm) [inline internal/backend/remote/backend.go:719]: **(idea)** We could log a warning here, or fail given that we don't know what did the "latest" mean for the backend when the workspace was created.

Comment (cam72cam) [reply internal/backend/remote/backend.go:719]: My preference here is warning if we want to take any action.

**Review by cam72cam — APPROVED** [review]

**Review by cube2222 — APPROVED** [review]
Other than the log message @cam72cam mentioned, LGTM!

File: internal/backend/remote/backend.go [hunk 3]
```diff
@@ -716,6 +709,28 @@ func (b *Remote) StateMgr(ctx context.Context, name string) (statemgr.Full, erro
...
+	}
+
+	// We trim the suffix if the backend workspace contains it.
+	wsv = version.Must(version.NewVersion(wsv)).Core().String()
```
Comment (Yantrio) [inline internal/backend/remote/backend.go:723]: In theory, we're trusting that the version returned here is either "latest" or a valid semver

By calling `Must` here will cause a panic if for some reason the remote backend has returned a non-valid, non-latest version.

In my opinion, I'd like to fail a little bit more gracefully and return the error and ensure that it's bubbled and handled higher up.

**Comment by kislerdm** [comment]
I bring PR back to "draft" state because setting the tofu version on init/workspace creation does not resolve the root cause of the issue. A logic regression introduced after the alpha3 release was identified. It leads to the deviation from the terraform behaviour: the [logic](https://github.com/opentofu/opentofu/blob/2d6dd8eda83b89b07a02eb4fdb2925fc448e83e7/internal/command/meta_backend.go#L1487) is not executed during the tofu run.

**Comment by kislerdm** [comment]
I close this PR because the issue which it addresses was resolved in #835.

**Pull request closed without merging**